lectures/jax_intro.md

```diff
@@ -0,0 +1,609 @@
+algebra operations and automatic differentiation.
+
+Loosely speaking, JAX is like NumPy with the addition of
+
+* automatic differentiation
+* automated GPU/TPU support
+* a just-in-time compiler
+
+One of the great benefits of JAX is that the same code can be run either on
+the CPU or on a hardware accelerator, such as a GPU or TPU.
+
+For example, JAX automatically builds and deploys kernels on the GPU whenever
+an accessible device is detected.
+
+### History
+
+In 2015, Google open-sourced part of its AI infrastructure called TensorFlow.
+
+Around two years later, Facebook open-sourced PyTorch beta, an alternative AI
+framework which is regarded as developer-friendly and more Pythonic than
+TensorFlow.
+
+By 2019, PyTorch was surging in popularity, adopted by Uber, Airbnb, Tesla and
+many other companies.
+
+In 2020, Google launched JAX as an open-source framework, simultaneously 
+beginning to shift away from TPUs to Nvidia GPUs. 
+
+In the last few years, uptake of Google JAX has accelerated rapidly, bringing
+attention back to Google-based machine learning architectures.
+
+
+### Installation
+
+JAX can be installed with or without GPU support by following [the install guide](https://github.com/google/jax).
+
+Note that JAX is pre-installed with GPU support on [Google Colab](https://colab.research.google.com/).
+
+If you do not have your own GPU, we recommend that you run this lecture on Colab.
+
++++
+
+## JAX as a NumPy Replacement
+
+
+One way to use JAX is as a plug-in NumPy replacement. Let's look at the similarities and differences.
+
+### Similarities
+
+
+The following import is standard, replacing `import numpy as np`:
+
+```{code-cell} ipython3
+import jax
+import jax.numpy as jnp
+```
+
+Now we can use `jnp` in place of `np` for the usual array operations:
+
+```{code-cell} ipython3
+a = jnp.asarray((1.0, 3.2, -1.5))
+```
+
+```{code-cell} ipython3
+print(a)
+```
+
+```{code-cell} ipython3
+print(jnp.sum(a))
+```
+
+```{code-cell} ipython3
+print(jnp.mean(a))
+```
+
+```{code-cell} ipython3
+print(jnp.dot(a, a))
+```
+
+However, the array object `a` is not a NumPy array:
+
+```{code-cell} ipython3
+a
+```
+
+```{code-cell} ipython3
+type(a)
+```
+
+Even scalar-valued maps on arrays return objects of type `DeviceArray`:
+
+```{code-cell} ipython3
+jnp.sum(a)
+```
+
+The term `Device` refers to the hardware accelerator (GPU or TPU), although JAX falls back to the CPU if no accelerator is detected.
+
+(In the terminology of GPUs, the "host" is the machine that launches GPU operations, while the "device" is the GPU itself.)
+
+Note that `DeviceArray` is a *future*; it allows Python to continue execution when the results of computation are not available immediately.
+
+So, Python can dispatch more jobs without waiting for the computation results to be returned by the device.
+
+This feature is called *asynchronous dispatch*, which hides Python overheads and reduces wait time.
```

Comment:
"So, " -> "This means that "

I suggest we put the above three lines in a note environment, since they are somewhat more technical than the other lines around them.